Posts/2023/06/Measuring-Download-Time.md

```diff
@@ -0,0 +1,544 @@
+---
+post_title: 'Measuring average download time'
+username: francisconabas
+categories: PowerShell
+tags: PowerShell, Automation, Performance, Measure-Command
+summary: This post shows how to measure average download time with PowerShell
+---
+
+One of the most overlooked roles of a systems administrator is to be able to troubleshoot
+network issues. How many times had you been in a situation where your servers are problematic,
+and someone asked you to check the network connectivity?
+One of the steps is checking downloading time, and speed, and although there are countless
+tools available, today we will learn how to do it natively, with PowerShell.
+
+## Methods
+
+We will focus on three methods, ranging from the easiest to the most complex, and discuss
+their pros and cons. These methods are the `Start-BitsTransfer` Cmdlet, using .NET with the
+`System.Net` namespace, and using the Windows native API.
+
+### `Start-BitsTransfer`
+
+BITS, or Background Intelligent Transfer Service is a Windows service that manages content transfer
+using HTTP or SMB. It was designed to manage the many aspects of content transfer, including cost,
+speed, priority, etc.
+For us, it also serves as an easy way of downloading files. Here is how you download a file from a
+web server using BITS:
+
+```powershell
+Start-BitsTransfer -Source 'https://www.myawesomeserver.com/Files/BitsDefinition.txt' `
+-Destination 'C:\BitsDefinition.txt'
+```
+
+Another great advantage of BITS is that it shows progress, which can be useful while
+downloading big files. In our case however, we want to know how long does it take to
+download a file. For this we will use a handy object of type `System.Diagnostics.Stopwatch`.
+
+```powershell
+$stopwatch = New-Object -TypeName 'System.Diagnostics.Stopwatch'
+
+$stopwatch.Start()
+Start-BitsTransfer -Source 'https://www.myawesomeserver.com/Files/BitsDefinition.txt' `
+-Destination 'C:\BitsDefinition.txt'
+$stopwatch.Stop()
+
+Write-Output $stopwatch.Elapsed
+```
+
+```powershell-console
+Days              : 0
+Hours             : 0
+Minutes           : 0
+Seconds           : 0
+Milliseconds      : 816
+Ticks             : 8165482
+TotalDays         : 9.45078935185185E-06
+TotalHours        : 0.000226818944444444
+TotalMinutes      : 0.0136091366666667
+TotalSeconds      : 0.8165482
+TotalMilliseconds : 816.5482
+```
+
+Awesome, we now have a baseline to build our script upon. First thing we will change is the file,
+since we are more interested on the speed, we can use temporary files to download, that also
```

Comment:
Breakup this run-on sentence.
```suggestion
Awesome, we now have a baseline to build our script upon. First thing we will change is the file.
Since we are more interested on the speed we can use temporary files to download. That also
```